### PR TITLE
perf(es/minifier): check whether a function is pure with O(1) cost

### PR DESCRIPTION
**Pull request by hyf0**
**Description:**

We might need to pass a giant pure function list to the SWC minifier as esbuild did [here](https://github.com/evanw/esbuild/blob/5fe21253ee75fb4c5ea395a0877b2a5ab51c3575/internal/config/globals.go).

The current implementation could slow the minifier.

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

**Comment by hyf0** [comment]
> Can we have those list also in swc minifier? (itself)

Absolute yes. I would do it in another PR. 

We might need to provide an option(enable by default) to disable it. Because users might do some hacky things to the environment.

**Comment by kdy1** [comment]
We have `pristine_globals`

**Comment by hyf0** [comment]
> We have `pristine_globals`

Get it.

**Comment by kdy1** [comment]
I think we can use a hash set of `(JsWord, JsWord)`.
Is there more complex expression other than those globals?

**Comment by hyf0** [comment]
> I think we can use a hash set of `(JsWord, JsWord)`. Is there more complex expression other than those globals?

Yes. Dealing `Expr::Member` and `Expr::Ident` should be enough base on the input [here](https://github.com/evanw/esbuild/blob/5fe21253ee75fb4c5ea395a0877b2a5ab51c3575/internal/config/globals.go).

File: crates/swc_ecma_minifier/src/metadata/mod.rs [hunk 1]
```diff
@@ -11,6 +14,37 @@ use swc_ecma_visit::{
 
 use crate::option::CompressOptions;
 
+#[derive(Debug, Eq)]
+struct HashEqIgnoreSpanExprRef<'a>(&'a Expr);
```
Comment (kdy1) [inline crates/swc_ecma_minifier/src/metadata/mod.rs:18]: I'm not sure if this will work.

Comment (kdy1) [reply crates/swc_ecma_minifier/src/metadata/mod.rs:18]: Can you add some tests?

File: crates/swc_ecma_minifier/src/metadata/mod.rs [hunk 2]
```diff
@@ -124,17 +167,18 @@ impl VisitMut for InfoMarker<'_> {
 
         if self.has_pure(n.span) {
             n.span = n.span.apply_mark(self.marks.pure);
-        } else if let Some(options) = self.options {
+        } else if let Some(pure_fns) = &self.pure_funcs {
             if let Callee::Expr(e) = &n.callee {
+                println!("--pure_fns: {pure_fns:#?}");
```
Comment (kdy1) [inline crates/swc_ecma_minifier/src/metadata/mod.rs:172]: Also, please remove this

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc_ecma_minifier/src/metadata/mod.rs (hunk 1 above)
Comment (hyf0) [reply crates/swc_ecma_minifier/src/metadata/mod.rs:18]: Sure. I guess we only need to test the `Eq` and `Hash` behaviors of `HashEqIgnoreSpanExprRef<'a>`.

Comment (kdy1) [reply crates/swc_ecma_minifier/src/metadata/mod.rs:18]: I think inserting and getting various values from `HashSet<HashEqIgnoreSpanExprRef>` is enough

File: crates/swc_ecma_minifier/src/metadata/tests.rs [hunk 3]
```diff
@@ -158,3 +158,63 @@
...
+    use swc_ecma_ast::*;
+
+    fn expr_ref(expr_ref: &Expr) -> HashEqIgnoreSpanExprRef {
+        HashEqIgnoreSpanExprRef(expr_ref)
+    }
+
+    testing::run_test(false, |_cm, _handler| {
+        let dummy_sp = DUMMY_SP;
+        let meaningful_sp = dummy_sp.apply_mark(Mark::new());
+
```
Comment (kdy1) [inline crates/swc_ecma_minifier/src/metadata/tests.rs:179]: Can we have some check for expressions which is not an ident nor a member expression?

Comment (hyf0) [reply crates/swc_ecma_minifier/src/metadata/tests.rs:179]: Done

File: crates/swc_ecma_minifier/src/metadata/mod.rs [hunk 4]
```diff
@@ -11,6 +14,37 @@ use swc_ecma_visit::{
...
+        // In pratice, most of cases/input we are dealing with are Expr::Member or
+        // Expr::Ident.
+        match self.0 {
+            Expr::Ident(i) => {
+                i.sym.hash(state);
+            }
+            Expr::Member(i) => {
+                Self(&i.obj).hash(state);
+                if let MemberProp::Ident(prop) = &i.prop {
+                    prop.sym.hash(state);
```
Comment (jridgewell) [inline crates/swc_ecma_minifier/src/metadata/mod.rs:37]: I think there's a bug here that `console.log` will hash the same as the identifier `consolelog`. We'd need to hash `"."` or similar sigil.

Comment (kdy1) [reply crates/swc_ecma_minifier/src/metadata/mod.rs:37]: Does it matter? `Eq` will be used if the hash value matches

Comment (jridgewell) [reply crates/swc_ecma_minifier/src/metadata/mod.rs:37]: Ohhhhhhh.

**Review by jridgewell — APPROVED** [review]

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_minifier

**Review by kdy1 — APPROVED** [review]

**Pull request merged**